### DISCUSSION TITLE
Buttons for Copy, Cut, Paste in Toolbar

### DISCUSSION THREAD

**Discussion by burnuser** [freeplane/freeplane#431] [Ideas]
Would be nice to have Buttons for Copy, Cut and Paste in the Toolbar like in (almost all) other programs for a quick workflow.

**Comment by rexelbartolome** [comment]
I'm curious for what reason you need a toolbar icon for these for? I do agree they're 'standard' but I feel like using keyboard shortcuts for those operations are already very familiar to the user (CTRL+X, CTRL+C and CTRL+V for Windows) that it's not 'necessary' for them to be in the toolbar, and it's also a much quicker workflow compared to clicking a button at the top of the screen.

The toolbar is currently undergoing a redesign and we're actually trying to cut down the amount of buttons there so if you could elaborate more on your use case that would be great

**Reply by burnuser** [reply]
At the moment, there are 3 possibilities for Copy, Cut and Paste of selected nodes, but all of them need multiple steps:
A.) Menu
1.) Klick on Menu
2.) Search for the Entry + go down with Cursor to the searched Entry
3.) Klick on Entry

B) Context Menu
1.) Right-click on Menu
2.) Search for the Entry (visually a bit longer, since the menu position is every time a bit different) + go down with Cursor to the searched Entry
3.) Klick on Entry

C) Keyboard shortcuts
1.) After rearranging nodes within the viewport using the mouse and my eyes on the screen, for Copy, Cut and Paste my hand must leave the mouse and my eyes must leave the screen.
2.) Short, but annoying and disturbing change to keyboard with eyes and both hands to enter the shortcut
3.) Change back: eyes to screen, hand to mouse

BUT:
With toolbar buttons for Copy, Cut and Paste it is always only 1 step, without any need to change position of hand(s) or focus of eyes!

And stating that fact, a "Delete node" Menu entry and Toolbar button is also missing!

And a little feedback for Your redesign project from my personal view as long-time Freeplane user and from my trainer experience:
In general: Rare used functions does not need a button. Often used functions should have one.

Could be removed:
I never use the Previous/Next map buttons. Each map has a Tab, that works well.
I never use Print map button. When I need such functions, I can use the menu.
I never use the Viewing/Editing button. Menu will do the job, if necessary.
I never use the (un)fold-buttons. Mousclick on the folding-point will do the job.

To keep:
Zoom
Undo/Redo
Find
New child/sibling (very important, especially for beginners!)
Options for node formatting (but clouds a used less often, could be removed to menu)

To add (for a distraction free mouse workflow):
Copy, Cut, Paste
Delete (node)
Add hyperlink (URL) ... at least for me an also very often needed function

If possible:
Make toolbar configurable: Selection, which buttons of the above should be shown.

And, by the way, two changed standard Preferences would be nice for better usability and accessibility:
1.) Behaviour - Selection method: "by click" (= Explicit user action gives an expectable explicit result. Actual standard "Delayed" is for newbies and people with handicaps much more confusing and harder to use!)
2.) Keyboard shortcut for "Insert new child node": was [TAB] in Freemind and would also be in Freeplane a much better solution than [Insert]. Because the [TAB] key is very prominent and easy to find and hit.
On the other hand, many keyboards (maybe only German ones, I don't know) have no explicit [Insert] Key at all. Only as alternative function of another key (often in the numeric keypad). Hard to find and bad to use for the maybe most important function in a map!.

**Reply by rexelbartolome** [reply]
Yes I agree, rarely used functions shouldn't be prioritized to have a visual button made for them.

Some of your points I agree with, like the `Insert` vs `Tab` and also `delete node`. My current setup is I rebound my `insert child node` to `Tab` then I rebound my `delete node` to `Escape`. Fortunately rebinding is an easy process, I'm not sure you already know

1. Go to Command Search `CTRL+F1`
2. Search for the command you want to rebind
3. While holding `CTRL` double click on the command and you'll be asked what button to bind it to

https://user-images.githubusercontent.com/88552647/173057346-21948f25-2f4b-4be0-886b-d91700d04fc9.mp4

`Add hyperlink (URL)` already has a keyboard shortcut but you can rebind it to something you're more comfortable using. Maybe one of the Function keys like `F4`

The `delete node` does have a Menu entry, can be found under `Insert Menu`

![image](https://user-images.githubusercontent.com/88552647/173058344-62ce2e4a-1116-44da-9916-d9417979d198.png)


For your case about the keyboard:

> C) Keyboard shortcuts
> 1.) After rearranging nodes within the viewport using the mouse and my eyes on the screen, for Copy, Cut and Paste my hand must leave the mouse and my eyes must leave the screen.
> 2.) Short, but annoying and disturbing change to keyboard with eyes and both hands to enter the shortcut
> 3.) Change back: eyes to screen, hand to mouse

I'm honestly not following what you're trying to say. The shortcuts I mentioned can all be done with one hand. 

![firefox_agr1mGXUFt](https://user-images.githubusercontent.com/88552647/173056084-ce741a73-ece6-49a1-ba40-4960121410c8.png)

Either the keyboard layout you have is vastly different from mine or you simply don't have enough muscle memory to do it with one hand without looking, which is also a fair point for your argument. Not everyone is that proficient with using a keyboard.

Maybe you can suggest the preference changes as a feature request for Dimitry. The `Insert` key I also agree with, I currently only have a 75% keyboard layout, so I was forced to rebind that command.

![image](https://user-images.githubusercontent.com/88552647/173059526-26989d3e-d0da-416e-b095-e72f282c0b6a.png)

And I too changed my behavior to `By Click` instead of `Delayed`

**Reply by burnuser** [reply]
Thank You!
Yes I know the method of rebinding key commands. It is on my standard To To list after a fresh Freeplane install.
But beside my one work I always think about the training situation when introducing newbies to Freeplane. 
And in many cases, people try my Freeplane recommendation on their own.
In every case, they should have the best work environment out oft the box, without the need of additional configurations or explanations, what to do after installation.
And that's why I recommend changing this two standards: Insert node with [TAB] and Selection "by click"

The latter has an additional effect.
People are trained to  expect some standard behaviours from computers. One of this standards is: If You want to select something, ... click on it!
And bringing new people to use Freeplane is much easier when they open Freeplane ... and feel like home! (Compared to other software.)
So, let's use industrial standards where ever possible:
1.) Selection by click
2. ) Toolbar-buttons for Copy, Cut, Paste
 + Delete (to close the mouse & button workflow!) ... and I never searched for this in the insert menu
 + Add Link (URL) => look and feel of known office software!

And yes, I know, some people like their keyboard very much. I don't want to underestimate or remove keyboard shortcuts.
But thinking in workflow terms, each forced (unnecessary) change between mouse and keyboard is - for normal users like me - a distraction of the workflow. (And yes, I often use two hands for shortcuts, ... like many other people I know.)
Additionally, people must know shortcuts, but everybody can discover standard buttons at a glance.
Make it easy for newbies to enter work with Freeplane! + Enable complete mouse-workflows - from creation to deletion of nodes - with toolbar buttons (additional to keyboard workflows).

And I think the best solution for toolbar buttons would really be a configuration screen for customization (like in a lot of other software). Nothing complicated, only checkmarks to show or not show existing and additional buttons (discussed and maybe some more, if you think).

**Reply by euu2021** [reply]
One idea, to improve the mouse centered workflow that you want, is to allow a different key binding to the middle mouse click. So, the user could use it for something convenient for him, like insert or delete node.

**Comment by EdoFro** [comment]
@rexelbartolome,
There have been many points discussed in this thread. I have no real new ideas to this, but I want to share my feedback for some topics so you have more opinions for your work.

> I never use the (un)fold-buttons. Mousclick on the folding-point will do the job.

I use them a lot. un/fold one level or all levels are commands that work different then clicking the unfolding symbol in the node, specially when working with big branches.

> To add (for a distraction free mouse workflow):
Copy, Cut, Paste
Delete (node)
Add hyperlink (URL) ... at least for me an also very often needed function

I don't need any of these, but I think maybe they should be easier to access in the pop up menu from the node (right click on the node). Maybe a first section with this standard commands.
I would add the Move command too.

For example to delete the selected node yoy have to navigate the popup menu. see following image:
![image](https://user-images.githubusercontent.com/35700575/173115235-553e22c3-9a3b-4267-9e93-8f7d5ab4092a.png)


> 1.) Behaviour - Selection method: "by click" (= Explicit user action gives an expectable explicit result. Actual standard "Delayed" is for newbies and people with handicaps much more confusing and harder to use!)

agree

> 2.) Keyboard shortcut for "Insert new child node": was [TAB] in Freemind and would also be in Freeplane a much better solution than [Insert]. Because the [TAB] key is very prominent and easy to find and hit.

agree

> On the other hand, many keyboards (maybe only German ones, I don't know) have no explicit [Insert] Key at all. Only as alternative function of another key (often in the numeric keypad). Hard to find and bad to use for the maybe most important function in a map!.

In my notebook to get *Insert*, I have to press [ Fn + Del ] . If I would use *Insert* for inserting a node, that will mean that I had to press
- Fn + Del --> insert new node
- Fn +shift + Del --> insert new parent node
- Fn +ctrl+ Del --> insert new summary node

This makes it not very user friendly and it must be used frequently. 
To make it worse, if I don't press Fn correctly, it deletes the selected node and all its descendants.

That's why I use F12 to insert new nodes. But I remember that as new user I wanted to use TAB to insert new nodes. That was what I felt natural at that moment.

### About Keyboard vs Mouse  workflows:

I think there are moments when using Freeplane that I want to work keyboard only and others when I want to go mouse only.
For example, 

- when typing new ideas and information, I work mostly keyboard only. I insert, delete, copy/paste, move nodes, apply styles, etc, using keyboard shortcuts and it works great. 
- when navigating and reorganizing the map, sometimes I work mouse only. For example when discussing with other people on how to organize a document or tasks in a project.
  -  it never is really a "mouse only" process, because there is allways something new to add (type), 
  - but sometimes you have to reach the keyboard only to delete or copy a node, and for those moments, a toolbar button or a popup command could be helpful

jm2c

edo

**Comment by dpolivaev** [comment]
I have added buttons for copy, paste, cut, and delete in the latest preview 1.10.3_10 because mouse lovers, toolbar users, and other freeplane newbies might miss them and because Rexel and I already put the effort into removing some other stuff from the toolbar.  I am not sure if we keep them or revert the change. Any feedback and considerations are welcome.

https://sourceforge.net/projects/freeplane/files/freeplane%20preview/

**Comment by dpolivaev** [comment]
Also in preview, 1.10.3_10 I changed the new node hot key to TAB instead of INSERT and changed the default selection method to "by click". Actually, both suggestions are not new and I think that I agree with them.
If you have an opinion please let us know.

**Reply by burnuser** [reply]
YES!
Much better now!

Only two remarks to make Delete/Remove function consistent in every way:

1.) In Popup/Context Menu, there is an existing "Remove" submenu.
That's where I would search for the command when not to find as first level command like Cut.

2.) Similar game in main menu. I would never expect and search "Remove node" as entry in submenu "Insert" - "New Node"
That's a classic "Edit" function which should also be placed there.
